src/NumericalAlgorithms/Integration/GslQuadAdaptive.hpp

```diff
@@ -0,0 +1,263 @@
+        gsl_integrand(std::forward<IntegrandType>(integrand)), lower_boundary,
+        upper_boundary, tolerance_abs, tolerance_rel, this->max_intervals_,
+        this->workspace_.get(), &result, &error);
+    detail::check_status_code(status_code);
+    return result;
+  }
+};
+
+/*!
+ * The algorithm for "IntegrableSingularitiesKnown" uses user-defined
+ * subintervals given by a vector of doubles `points`, where each element
+ * denotes an interval boundary.
+ */
+template <>
+class GslQuadAdaptive<GslIntegralType::IntegrableSingularitiesKnown>
+    : public detail::GslQuadAdaptiveImpl {
+ public:
+  using detail::GslQuadAdaptiveImpl::GslQuadAdaptiveImpl;
+  template <typename IntegrandType>
+  double operator()(IntegrandType&& integrand, std::vector<double> points,
+                    const double tolerance_abs,
+                    const double tolerance_rel = 0.) const noexcept {
+    double result;
+    double error;
+    detail::disable_gsl_error_handling();
+    const int status_code = gsl_integration_qagp(
+        gsl_integrand(std::forward<IntegrandType>(integrand)), points.data(),
+        points.size(), tolerance_abs, tolerance_rel, this->max_intervals_,
+        this->workspace_.get(), &result, &error);
+    detail::check_status_code(status_code);
+    return result;
+  }
+};
+
+/*!
+ * The algorithm for "InfiniteInterval" uses the semi-open interval
+ * \f$ (0, 1] \f$ to map to an infinite interval \f$ (-\infty, +\infty) \f$. Its
+ * function takes no parameters other than limits `tolerance_abs` and optional
+ * `tolerance_rel` for the absolute error and the relative error.
+ */
+template <>
+class GslQuadAdaptive<GslIntegralType::InfiniteInterval>
+    : public detail::GslQuadAdaptiveImpl {
+ public:
+  using detail::GslQuadAdaptiveImpl::GslQuadAdaptiveImpl;
+  template <typename IntegrandType>
+  double operator()(IntegrandType&& integrand, const double tolerance_abs,
+                    const double tolerance_rel = 0.) noexcept {
+    double result;
+    double error;
+    detail::disable_gsl_error_handling();
+    const int status_code = gsl_integration_qagi(
+        gsl_integrand(std::forward<IntegrandType>(integrand)), tolerance_abs,
+        tolerance_rel, this->max_intervals_, this->workspace_.get(), &result,
+        &error);
+    detail::check_status_code(status_code);
+    return result;
+  }
+};
+
+/*!
+ * The algorithm for "UpperBoundaryInfinite" maps the semi-open interval
+ * \f$ (0, 1] \f$ to a semi-infinite interval \f$ [a, +\infty) \f$, where a is
+ * given by `lower_boundary`.
+ */
+template <>
+class GslQuadAdaptive<GslIntegralType::UpperBoundaryInfinite>
+    : public detail::GslQuadAdaptiveImpl {
+ public:
+  using detail::GslQuadAdaptiveImpl::GslQuadAdaptiveImpl;
+  template <typename IntegrandType>
+  double operator()(IntegrandType&& integrand, const double lower_boundary,
+                    const double tolerance_abs,
+                    const double tolerance_rel = 0.) noexcept {
```

Comment:
you inconsistently mark some of these member functions `const`, or `noexcept`. It looks like they should all be `const`